sdk/core/azure_core/src/task/standard_spawn.rs

```diff
@@ -30,39 +27,54 @@ struct ThreadJoinFuture {
 #[cfg(not(target_arch = "wasm32"))]
 #[derive(Default)]
 struct ThreadJoinState {
-    join_handle: Option<thread::JoinHandle<()>>,
+    join_handle:
+        Option<thread::JoinHandle<std::result::Result<(), Box<dyn std::error::Error + Send>>>>,
     waker: Option<Waker>,
+    thread_finished: bool,
 }
 
 #[cfg(not(target_arch = "wasm32"))]
 impl Future for ThreadJoinFuture {
-    type Output = std::result::Result<(), Box<dyn std::error::Error>>;
+    type Output = std::result::Result<(), Box<dyn std::error::Error + Send>>;
 
-    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
-        // We need to check if the thread is done
-        // Since there's no non-blocking way to check thread completion in std,
-        // we attempt to join only once during polling
-        let mut join_state = self
-            .join_state
-            .lock()
-            .map_err(|e| format!("Failed to lock join state: {:?}", e))?;
-        if let Some(handle) = &join_state.join_handle {
-            if handle.is_finished() {
-                // Since we know the thread is finished, we can safely take the handle
-                // and join it without blocking.
-                match join_state.join_handle.take().unwrap().join() {
-                    Ok(_) => Poll::Ready(Ok(())),
-                    Err(e) => Poll::Ready(Err(format!("Thread panicked: {:?}", e).into())),
-                }
-            } else {
-                // Thread is still running, so we need to register the waker
-                // for when it completes.
-                join_state.waker = Some(_cx.waker().clone());
-                Poll::Pending
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut join_state = self.join_state.lock().map_err(|e| {
+            debug!("Failed to lock join state: {}", e);
+            Box::new(crate::Error::message(
+                crate::error::ErrorKind::Other,
+                format!("Thread panicked: {:?}", e),
+            )) as Box<dyn std::error::Error + Send>
```

Comment:
Tip: you shouldn't need to box the error yourself. I've never once had to when functions return a `Result<T, Box<dyn Error>`.

```rust
fn example() -> Result<(), Box<dyn std::error::Error>> {
    Err("error".into())
}

fn main() {
    match example() {
        Ok(()) => {},
        Err(err) => eprintln!("{err:?}"),
    }
}
```